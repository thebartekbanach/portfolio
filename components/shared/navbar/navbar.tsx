import { FC } from "react";
import { Logo } from "./logo";
import { NavItem, NavItemsWrapper, NavItems } from "./styles/navbarItems";
import { NavbarWrapper } from "./styles/navbarWrapper";
import {
	MobileMenuToggleButton,
	MobileMenuToggleButtonIcon
} from "./styles/mobileMenuToggleButton";
import { ContactButton } from "./styles/contactButton";

interface NavigationItem {
	id: string;
	url: string;
	content: string;
}

interface NavItemsListProps {
	items: NavigationItem[];
	onNavItemClick: (id: string) => void;
}

export const NavItemsList: FC<NavItemsListProps> = ({ items, onNavItemClick }) => (
	<>
		{items.map(item => (
			<NavItem key={item.id} onClick={() => onNavItemClick(item.id)}>
				<a href={item.url}>{item.content}</a>
			</NavItem>
		))}
	</>
);

interface NavigationProps extends NavItemsListProps {
	mobileMenuState: boolean;
	logoDirectoryName: string;
	logoUserName: string;
	contactButtonText: string;
	onContactButtonClick: () => void;
	onMobileMenuToggleButtonClick: () => void;
}

export const Navbar: FC<NavigationProps> = ({
	logoDirectoryName,
	logoUserName,
	contactButtonText,
	onContactButtonClick,
	mobileMenuState,
	onMobileMenuToggleButtonClick,
	...navItemsListProps
}) => (
	<NavbarWrapper>
		<Logo directoryName={logoDirectoryName} userName={logoUserName} />
		<MobileMenuToggleButton
			mobileMenuState={mobileMenuState}
			onClick={onMobileMenuToggleButtonClick}
		>
			<MobileMenuToggleButtonIcon mobileMenuState={mobileMenuState} />
		</MobileMenuToggleButton>
		<NavItemsWrapper isShownOnMobile={mobileMenuState}>
			<NavItems isShownOnMobile={mobileMenuState}>
				<NavItemsList {...navItemsListProps} />
				<NavItem key="contact-button">
					<ContactButton onClick={onContactButtonClick}>
						{contactButtonText}
					</ContactButton>
				</NavItem>
			</NavItems>
		</NavItemsWrapper>
	</NavbarWrapper>
);
