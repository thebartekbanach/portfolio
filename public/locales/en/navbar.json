{
	"logo": {
		"rootPath": "/dev/",
		"userName": "BartekBanach"
	},
	"navbarItems": [
		{
			"id": 1,
			"url": "#skills",
			"content": "Skills"
		},
		{
			"id": 2,
			"url": "#realizations",
			"content": "Realizations"
		},
		{
			"id": 3,
			"url": "#contact",
			"content": "Contact"
		}
	],
	"contactButtonContent": "Write to me"
}
