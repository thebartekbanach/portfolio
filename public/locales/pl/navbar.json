{
	"logo": {
		"rootPath": "/dev/",
		"userName": "BartekBanach"
	},
	"navbarItems": [
		{
			"id": 1,
			"url": "#umiejetnosci",
			"content": "Umiejętności"
		},
		{
			"id": 2,
			"url": "#realizacje",
			"content": "Realizacje"
		},
		{
			"id": 3,
			"url": "#kontakt",
			"content": "Kontakt"
		}
	],
	"contactButtonContent": "Napisz do mnie"
}
