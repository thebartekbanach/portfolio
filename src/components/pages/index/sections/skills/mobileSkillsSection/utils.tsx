import debounce from "lodash.debounce";

const scrollBy = (top: number, abortAfterTimeMs = 1500) =>
	new Promise((resolve, reject) => {
		const abortTimeoutId = setTimeout(() => {
			window.removeEventListener("scroll", scrollEventHandler, eventOptions);
			reject("scrollBy aborted due to scroll end timeout");
		}, abortAfterTimeMs);

		const scrollEventHandler = debounce(() => {
			clearTimeout(abortTimeoutId);
			resolve();
		}, 100);

		const eventOptions: AddEventListenerOptions = { passive: true };

		window.addEventListener("scroll", scrollEventHandler, eventOptions);

		window.scrollBy({
			behavior: "smooth",
			top
		});
	});

export class StickGuard {
	private target: (EventTarget & Element) | null = null;
	private offsetTop = 0;

	private guardScrollPosition = () => {
		if (this.target === null) {
			return;
		}

		const position = this.target.getBoundingClientRect().top - this.offsetTop;
		window.scrollBy({ left: 0, top: position });

		window.requestAnimationFrame(this.guardScrollPosition);
	};

	stickTo(target: EventTarget & Element, offsetTop = 0) {
		this.target = target;
		this.offsetTop = offsetTop;

		this.guardScrollPosition();
	}

	async scrollToAndStick(target: EventTarget & Element, offsetTop = 0) {
		const position = target.getBoundingClientRect().top - offsetTop;

		if (position > 1 || position < -1) {
			await scrollBy(position);
		}

		this.stickTo(target, offsetTop);
	}

	stopSticking() {
		this.target = null;
	}
}
