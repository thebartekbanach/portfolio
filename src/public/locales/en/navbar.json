{
	"historyEntry": "{{sectionName}} - Bartek Banach portfolio",
	"logo": {
		"href": "/#start",
		"rootPath": "/dev/",
		"userName": "BartekBanach"
	},
	"navbarItems": [
		{
			"href": "/#skills",
			"text": "Skills"
		},
		{
			"href": "/#realizations",
			"text": "Realizations"
		},
		{
			"href": "/#contact",
			"text": "Contact"
		}
	],
	"contactButton": {
		"href": "/#contact",
		"text": "Write to me"
	}
}
