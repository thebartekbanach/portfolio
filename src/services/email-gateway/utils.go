package main

import (
	"io/ioutil"
	"log"
	"mime"
	"strings"

	"github.com/tidwall/gjson"
)

// translation files path "/root/path/to/translations/${lang}/translation.json"
func convertSubjectIDToTranslatedString(subjectID, lang, translationFilesPath, jsonTranslationPath string) string {
	translationFilePath := strings.ReplaceAll(translationFilesPath, "{lang}", lang)

	translation, err := ioutil.ReadFile(translationFilePath)
	if err != nil {
		log.Panicln("Convert subjectID to translated string error: ", err)
	}

	query := strings.ReplaceAll(jsonTranslationPath, "{id}", subjectID)
	result := gjson.Get(string(translation), query)

	encodedString := mime.QEncoding.Encode("utf-8", result.String())

	return encodedString
}
