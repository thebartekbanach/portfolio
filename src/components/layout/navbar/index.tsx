import React, { FC, useState } from "react";

import { useTranslation } from "~/utils/i18next";

import { Logo } from "./logo";
import { NavItems } from "./navItems";
import { NavbarContainer } from "./styles";

interface NavbarProps {
	pageSubPath?: string;
}

export const Navbar: FC<NavbarProps> = ({ pageSubPath }) => {
	const [t] = useTranslation("navbar");
	const [mobileMenuOpenState, setMobileMenuOpenState] = useState(false);

	const toggleMenu = () => setMobileMenuOpenState(!mobileMenuOpenState);

	const scrollToSectionOrRedirect = (e: React.MouseEvent<HTMLAnchorElement, MouseEvent>) => {
		const sectionUrl = e.currentTarget.getAttribute("href");
		const sectionName = e.currentTarget.getAttribute("data-section-name");
		const locationWithSelector = window.location.pathname + "#";

		if (!sectionUrl?.startsWith(locationWithSelector)) {
			return;
		}

		e.preventDefault();

		setMobileMenuOpenState(false);

		const sectionSelector = sectionUrl.slice(locationWithSelector.length - 1);
		const section = document.querySelector(sectionSelector);

		window.history.pushState(
			null,
			t("historyEntry").replaceAll("{{sectionName}}", sectionName ?? ""),
			sectionUrl
		);

		if (section === null) {
			throw new Error(
				`Cannot smoothly scroll to section that does not exist on current page: ${sectionSelector}`
			);
		}

		section?.scrollIntoView({ behavior: "smooth" });
	};

	const scrollToTopOfPageOrRedirect = (e: React.MouseEvent<HTMLAnchorElement, MouseEvent>) => {
		const logoHrefUrl = e.currentTarget.getAttribute("href");
		const hrefWithoutSection = logoHrefUrl?.slice(0, logoHrefUrl.indexOf("#"));

		if (window.location.pathname !== hrefWithoutSection) {
			return;
		}

		e.preventDefault();

		setMobileMenuOpenState(false);

		window.history.pushState(null, t("pageMeta:page.rootTitle"), hrefWithoutSection);
	};

	return (
		<NavbarContainer>
			<Logo subPage={pageSubPath} onLogoClick={scrollToTopOfPageOrRedirect} />
			<NavItems
				items={t("navbarItems", { returnObjects: true })}
				contactButton={t("contactButton", { returnObjects: true })}
				isMobileMenuOpen={mobileMenuOpenState}
				onMobileMenuToggleButtonClick={toggleMenu}
				onNavItemClick={scrollToSectionOrRedirect}
			/>
		</NavbarContainer>
	);
};
