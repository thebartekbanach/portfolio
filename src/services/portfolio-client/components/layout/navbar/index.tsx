import React, { FC, useCallback, useEffect, useState } from "react";

import { useTranslation } from "~/utils/i18next";
import { scrollToElement } from "~/utils/scrollToElement";

import { Logo } from "./logo";
import { NavItems } from "./navItems";
import { NavbarContainer } from "./styles";

interface NavbarProps {
	pageSubPath?: string;
}

export const Navbar: FC<NavbarProps> = ({ pageSubPath }) => {
	const [t] = useTranslation("navbar");
	const [mobileMenuOpenState, setMobileMenuOpenState] = useState(false);

	const closeOnScrollListener = useCallback(() => {
		setMobileMenuOpenState(false);
	}, []);

	useEffect(() => {
		if (mobileMenuOpenState === true) {
			const scrollListenerConfig = { once: true, passive: true };
			document.addEventListener("scroll", closeOnScrollListener, scrollListenerConfig);
		}
	}, [mobileMenuOpenState]);

	const toggleMenu = () => {
		document.removeEventListener("scroll", closeOnScrollListener);
		setMobileMenuOpenState(!mobileMenuOpenState);
	};

	const scrollToSectionOrRedirect = (e: React.MouseEvent<HTMLAnchorElement, MouseEvent>) => {
		const sectionUrl = e.currentTarget.getAttribute("href");
		const sectionName = e.currentTarget.getAttribute("data-section-name");
		const locationWithSelector = window.location.pathname + "#";

		if (!sectionUrl?.startsWith(locationWithSelector)) {
			return;
		}

		e.preventDefault();

		setMobileMenuOpenState(false);

		const sectionSelector = sectionUrl.slice(locationWithSelector.length - 1);
		const section = document.querySelector(sectionSelector);

		window.history.pushState(
			null,
			t("historyEntry").replaceAll("{{sectionName}}", sectionName ?? ""),
			sectionUrl
		);

		if (section === null) {
			throw new Error(
				`Cannot smoothly scroll to section that does not exist on current page: ${sectionSelector}`
			);
		}

		scrollToElement(section);
	};

	const scrollToTopOfPageOrRedirect = (e: React.MouseEvent<HTMLAnchorElement, MouseEvent>) => {
		const logoHrefUrl = e.currentTarget.getAttribute("href");
		const hrefWithoutSection = logoHrefUrl?.slice(0, logoHrefUrl.indexOf("#"));

		if (window.location.pathname !== hrefWithoutSection) {
			return;
		}

		e.preventDefault();

		setMobileMenuOpenState(false);

		window.history.pushState(null, t("pageMeta:page.rootTitle"), hrefWithoutSection);
	};

	return (
		<NavbarContainer>
			<Logo subPage={pageSubPath} onLogoClick={scrollToTopOfPageOrRedirect} />
			<NavItems
				items={t("navbarItems", { returnObjects: true })}
				contactButton={t("contactButton", { returnObjects: true })}
				isMobileMenuOpen={mobileMenuOpenState}
				onMobileMenuToggleButtonClick={toggleMenu}
				onNavItemClick={scrollToSectionOrRedirect}
			/>
		</NavbarContainer>
	);
};
