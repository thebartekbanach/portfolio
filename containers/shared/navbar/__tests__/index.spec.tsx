jest.mock("~/utils/i18next");

import { Navbar } from "~/containers/shared/navbar";
import { useTranslationMock } from "~/tests/utils/useTranslationMock";
import { mountWithRedux, FakeStore } from "~/tests/utils/mountWithRedux";

describe(Navbar, () => {
	it("should match snapshot", () => {
		useTranslationMock("navbar", {
			"logo.rootPath": "/dev/",
			"logo.userName": "BartekBanach",
			navbarItems: [
				{
					id: 1,
					url: "#skills",
					content: "Skills"
				},
				{
					id: 2,
					url: "#realizations",
					content: "Realizations"
				},
				{
					id: 3,
					url: "#contact",
					content: "Contact"
				}
			],
			contactButtonContent: "Write to me"
		});

		const store = new FakeStore({
			navbar: {
				mobileNavbarOpenState: false
			}
		});

		const result = mountWithRedux(store, <Navbar />);

		expect(result).toMatchSnapshot();
	});

	it("should render correctly", () => {});
});
