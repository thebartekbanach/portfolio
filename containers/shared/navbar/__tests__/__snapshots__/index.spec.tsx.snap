// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar should match snapshot 1`] = `
<Provider
  store={
    FakeStore {
      "dispatch": [Function],
      "dispatchedActions": Array [],
      "getState": [Function],
      "replaceReducer": [Function],
      "state": Object {
        "navbar": Object {
          "mobileNavbarOpenState": false,
        },
      },
      "subscribe": [Function],
      "subscribers": Array [
        [Function],
      ],
      Symbol(observable): [Function],
    }
  }
>
  <Navbar>
    <Navbar
      contactButtonText="Write to me"
      items={
        Array [
          Object {
            "content": "Skills",
            "id": 1,
            "url": "#skills",
          },
          Object {
            "content": "Realizations",
            "id": 2,
            "url": "#realizations",
          },
          Object {
            "content": "Contact",
            "id": 3,
            "url": "#contact",
          },
        ]
      }
      logoDirectoryName="/dev/"
      logoUserName="BartekBanach"
      mobileMenuState={false}
      onContactButtonClick={[Function]}
      onMobileMenuToggleButtonClick={[Function]}
      onNavItemClick={[Function]}
    >
      <navbarWrapper__NavbarWrapper>
        <div
          className="navbarWrapper__NavbarWrapper-sc-17lj7q5-0 pezyW"
        >
          <Logo
            directoryName="/dev/"
            userName="BartekBanach"
          >
            <logo__LogoWrapper>
              <div
                className="logo__LogoWrapper-sc-1xqfz6q-3 ddKwIZ"
              >
                <logo__LogoElement>
                  <span
                    className="logo__LogoElement-sc-1xqfz6q-2 gtrWvO"
                  >
                    <logo__DirectoryName>
                      <span
                        className="logo__DirectoryName-sc-1xqfz6q-0 dDMeVY"
                      >
                        /dev/
                      </span>
                    </logo__DirectoryName>
                    <logo__UserName>
                      <span
                        className="logo__UserName-sc-1xqfz6q-1 fJaYBw"
                      >
                        BartekBanach
                      </span>
                    </logo__UserName>
                  </span>
                </logo__LogoElement>
              </div>
            </logo__LogoWrapper>
          </Logo>
          <mobileMenuToggleButton__MobileMenuToggleButton
            mobileMenuState={false}
            onClick={[Function]}
          >
            <button
              className="mobileMenuToggleButton__MobileMenuToggleButton-sc-5lyury-1 kZVuFn"
              onClick={[Function]}
            >
              <mobileMenuToggleButton__MobileMenuToggleButtonIcon
                mobileMenuState={false}
              >
                <div
                  className="mobileMenuToggleButton__MobileMenuToggleButtonIcon-sc-5lyury-0 dFPhIK"
                />
              </mobileMenuToggleButton__MobileMenuToggleButtonIcon>
            </button>
          </mobileMenuToggleButton__MobileMenuToggleButton>
          <navbarItems__NavItemsWrapper
            isShownOnMobile={false}
          >
            <nav
              className="navbarItems__NavItemsWrapper-sc-3w6yz4-2 kWuMHv"
            >
              <navbarItems__NavItems
                isShownOnMobile={false}
              >
                <ul
                  className="navbarItems__NavItems-sc-3w6yz4-1 ecgEsl"
                >
                  <NavItemsList
                    items={
                      Array [
                        Object {
                          "content": "Skills",
                          "id": 1,
                          "url": "#skills",
                        },
                        Object {
                          "content": "Realizations",
                          "id": 2,
                          "url": "#realizations",
                        },
                        Object {
                          "content": "Contact",
                          "id": 3,
                          "url": "#contact",
                        },
                      ]
                    }
                    onNavItemClick={[Function]}
                  >
                    <navbarItems__NavItem
                      key="1"
                      onClick={[Function]}
                    >
                      <li
                        className="navbarItems__NavItem-sc-3w6yz4-0 irozit"
                        onClick={[Function]}
                      >
                        <a
                          href="#skills"
                        >
                          Skills
                        </a>
                      </li>
                    </navbarItems__NavItem>
                    <navbarItems__NavItem
                      key="2"
                      onClick={[Function]}
                    >
                      <li
                        className="navbarItems__NavItem-sc-3w6yz4-0 irozit"
                        onClick={[Function]}
                      >
                        <a
                          href="#realizations"
                        >
                          Realizations
                        </a>
                      </li>
                    </navbarItems__NavItem>
                    <navbarItems__NavItem
                      key="3"
                      onClick={[Function]}
                    >
                      <li
                        className="navbarItems__NavItem-sc-3w6yz4-0 irozit"
                        onClick={[Function]}
                      >
                        <a
                          href="#contact"
                        >
                          Contact
                        </a>
                      </li>
                    </navbarItems__NavItem>
                  </NavItemsList>
                  <navbarItems__NavItem
                    key="contact-button"
                  >
                    <li
                      className="navbarItems__NavItem-sc-3w6yz4-0 irozit"
                    >
                      <contactButton__ContactButton
                        onClick={[Function]}
                      >
                        <button
                          className="contactButton__ContactButton-sc-64ke0n-0 hooFnr"
                          onClick={[Function]}
                        >
                          Write to me
                        </button>
                      </contactButton__ContactButton>
                    </li>
                  </navbarItems__NavItem>
                </ul>
              </navbarItems__NavItems>
            </nav>
          </navbarItems__NavItemsWrapper>
        </div>
      </navbarWrapper__NavbarWrapper>
    </Navbar>
  </Navbar>
</Provider>
`;
